jetty-security/src/main/java/org/eclipse/jetty/security/authentication/ConfigurableSpnegoAuthenticator.java

```diff
@@ -0,0 +1,232 @@
+ *
+ * @see org.eclipse.jetty.security.ConfigurableSpnegoLoginService
+ */
+public class ConfigurableSpnegoAuthenticator extends LoginAuthenticator
+{
+    private static final Logger LOG = Log.getLogger(ConfigurableSpnegoAuthenticator.class);
+
+    private final String _authMethod;
+    private Duration _authenticationDuration = Duration.ofNanos(-1);
+
+    public ConfigurableSpnegoAuthenticator()
+    {
+        this(Constraint.__SPNEGO_AUTH);
+    }
+
+    /**
+     * Allow for a custom authMethod value to be set for instances where SPNEGO may not be appropriate
+     *
+     * @param authMethod the auth method
+     */
+    public ConfigurableSpnegoAuthenticator(String authMethod)
+    {
+        _authMethod = authMethod;
+    }
+
+    @Override
+    public String getAuthMethod()
+    {
+        return _authMethod;
+    }
+
+    /**
+     * @return the authentication duration
+     */
+    public Duration getAuthenticationDuration()
+    {
+        return _authenticationDuration;
+    }
+
+    /**
+     * <p>Sets the duration of the authentication.</p>
+     * <p>A negative duration means that the authentication is only valid for the current request.</p>
+     * <p>A zero duration means that the authentication is valid forever.</p>
+     * <p>A positive value means that the authentication is valid for the specified duration.</p>
+     *
+     * @param authenticationDuration the authentication duration
+     */
+    public void setAuthenticationDuration(Duration authenticationDuration)
+    {
+        _authenticationDuration = authenticationDuration;
+    }
+
+    @Override
+    public Authentication validateRequest(ServletRequest req, ServletResponse res, boolean mandatory) throws ServerAuthException
+    {
+        if (!mandatory)
+            return new DeferredAuthentication(this);
+
+        HttpServletRequest request = (HttpServletRequest)req;
+        HttpServletResponse response = (HttpServletResponse)res;
+
+        HttpSession httpSession = request.getSession(false);
+        if (httpSession != null)
+        {
+            UserIdentityHolder holder = (UserIdentityHolder)httpSession.getAttribute(UserIdentityHolder.ATTRIBUTE);
+            if (holder != null)
+            {
+                UserIdentity identity = holder._userIdentity;
+                if (identity != null)
+                {
+                    Duration authnDuration = getAuthenticationDuration();
+                    if (!authnDuration.isNegative())
+                    {
+                        boolean expired = !authnDuration.isZero() && Instant.now().isAfter(holder._validFrom.plus(authnDuration));
+                        // Allow non-GET requests even if they're expired, so that
+                        // the client does not need to send the request content again.
+                        if (!expired || !HttpMethod.GET.is(request.getMethod()))
+                            return new UserAuthentication(getAuthMethod(), identity);
```

Comment:
Should a token be set on the response if we are not deferred (same as 149 below?)

Reply:
We should not go through GSS context establishment here, because we trust the HTTP session.
Here, the request may not even have the client token, so we don't send the server token back.

However, if the request does have a token, then perhaps we should really skip the HTTP session check and run the GSS context.